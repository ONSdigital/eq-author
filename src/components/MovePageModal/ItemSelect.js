import React from "react";
import styled from "styled-components";
import { colors } from "constants/theme";
import PropTypes from "prop-types";
import VisuallyHidden from "components/VisuallyHidden";
import iconMoveIndicator from "./icon-move-indicator.svg";
import { uniqueId } from "lodash";
import withChangeHandler from "../Forms/withChangeHandler";

const Input = VisuallyHidden.withComponent("input");
Input.defaultProps = {
  type: "radio"
};

const Truncated = styled.span`
  display: block;
  white-space: nowrap;
  overflow: hidden;
  text-overflow: ellipsis;
`;

const Label = styled.label`
  display: flex;
  align-items: center;
  padding: 0.5em 0.5em 0.5em 2.25em;
  position: relative;

  &:hover {
    background-color: ${colors.lighterGrey};
  }

  ${/* sc-sel */ Input}:focus + & {
    outline: 2px solid ${colors.text};
    outline-offset: -2px;
  }

  ${/* sc-sel */ Input}:checked + & {
    background-color: ${colors.blue};
    color: ${colors.white};
  }

  &:hover::before {
    background-color: ${colors.text};
  }

  &:hover::before,
  ${/* sc-sel */ Input}:checked + &::before {
    content: "";
    position: absolute;
    left: 1rem;
    width: 1em;
    height: 1em;
    mask-image: url('${iconMoveIndicator}');
    mask-repeat: no-repeat;
    mask-position: center center;
  }

  ${/* sc-sel */ Input}:checked + &::before {
    background-color: white;
  }
`;

export const Option = ({
  name,
  selected,
  value,
  onChange,
  id = uniqueId("ItemList_Option"),
  children,
  ...otherProps
}) => (
  <div {...otherProps}>
    <Input
      value={value}
      id={id}
      onChange={onChange}
      checked={selected}
      name={name}
    />

    <Label selected={selected} htmlFor={id}>
      <Truncated>{children}</Truncated>
    </Label>
  </div>
);

Option.propTypes = {
  selected: PropTypes.bool,
  value: PropTypes.string.isRequired,
  name: PropTypes.string,
  onChange: PropTypes.func,
  id: PropTypes.string,
  children: PropTypes.node.isRequired
};

const ItemSelect = ({ children, value, name, onChange, ...otherProps }) => (
  <div {...otherProps}>
    {React.Children.map(children, child =>
      React.cloneElement(child, {
        onChange,
        name,
        selected: child.props.value === value
      })
    )}
  </div>
);

ItemSelect.propTypes = {
  children: PropTypes.node.isRequired,
  value: PropTypes.string.isRequired,
  name: PropTypes.string.isRequired,
  onChange: PropTypes.func.isRequired
};

export default withChangeHandler(ItemSelect);
