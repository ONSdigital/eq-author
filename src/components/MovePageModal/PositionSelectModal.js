import React from "react";
import PropTypes from "prop-types";
import CustomPropTypes from "custom-prop-types";
import styled from "styled-components";
import Button from "components/Button";
import ItemSelect, { Option } from "./ItemSelect";
import PlainModal, { Title, Body, Fieldset } from "./PlainModal";
import ButtonGroup from "components/ButtonGroup";
import getTextFromHTML from "utils/getTextFromHTML";

const StyledButtonGroup = styled(ButtonGroup).attrs({
  horizontal: true,
  align: "right"
})`
  padding: 1em;
  justify-content: space-around;
  border-top: 1px solid #aeaeae;
`;

const StyledButton = styled(Button)`
  flex: 1;
`;

const PositionSelectModal = ({
  isOpen,
  onClose,
  onChange,
  selectedPosition,
  pages,
  onConfirm
}) => (
  <PlainModal isOpen={isOpen} onClose={onClose}>
    <form onSubmit={onConfirm}>
      <Fieldset>
        <Title>Position</Title>
        <Body>
          <ItemSelect
            value={String(selectedPosition)}
            onChange={onChange}
            name="position"
          >
            {pages.map((page, i) => (
              <Option key={i} value={String(i)}>
                {getTextFromHTML(page.title)}
              </Option>
            ))}
          </ItemSelect>
        </Body>
      </Fieldset>
      <StyledButtonGroup>
        <StyledButton onClick={onClose} secondary type="button">
          Cancel
        </StyledButton>
        <StyledButton type="submit" primary>
          Move
        </StyledButton>
      </StyledButtonGroup>
    </form>
  </PlainModal>
);

PositionSelectModal.propTypes = {
  isOpen: PropTypes.bool.isRequired,
  onClose: PropTypes.func.isRequired,
  onChange: PropTypes.func.isRequired,
  selectedPosition: PropTypes.number.isRequired,
  pages: PropTypes.arrayOf(CustomPropTypes.page).isRequired,
  onConfirm: PropTypes.func.isRequired
};

export default PositionSelectModal;
