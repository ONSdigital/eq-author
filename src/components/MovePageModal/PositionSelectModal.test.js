import React from "react";
import PositionSelectModal from "./PositionSelectModal";
import { shallow } from "enzyme";
import ItemSelect from "./ItemSelect";

describe("MovePageModal/PositionSelectModal", () => {
  const pages = [{ id: "1", title: "page 1" }, { id: "2", title: "page 2" }];

  const createWrapper = (props = {}, render = shallow) =>
    render(
      <PositionSelectModal
        isOpen
        onClose={jest.fn()}
        pages={pages}
        selectedPosition={0}
        onChange={jest.fn()}
        onConfirm={jest.fn()}
        {...props}
      />
    );

  it("should render", () => {
    expect(createWrapper()).toMatchSnapshot();
  });

  it("should call onChange when pages position is selected", () => {
    const onChange = jest.fn();
    const wrapper = createWrapper({ onChange });

    wrapper.find(ItemSelect).simulate("change");
    expect(onChange).toHaveBeenCalled();
  });

  it("should call onConfirm when submit button is clicked", () => {
    const onConfirm = jest.fn();
    const wrapper = createWrapper({ onConfirm });

    wrapper.find(`form`).simulate("submit");
    expect(onConfirm).toHaveBeenCalled();
  });

  it("should close the modal when cancel button is clicked", () => {
    const onClose = jest.fn();
    const wrapper = createWrapper({ onClose });

    wrapper.find(`[type="button"]`).simulate("click");
    expect(onClose).toHaveBeenCalled();
  });
});
