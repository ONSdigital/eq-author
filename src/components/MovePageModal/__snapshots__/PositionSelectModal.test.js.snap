// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MovePageModal/PositionSelectModal should render 1`] = `
<PlainModal
  isOpen={true}
  onClose={[Function]}
>
  <form
    onSubmit={[Function]}
  >
    <PlainModal__Fieldset>
      <PlainModal__Title>
        Position
      </PlainModal__Title>
      <Body>
        <withChangeHandler(ItemSelect)
          name="position"
          onChange={[Function]}
          value="0"
        >
          <Option
            key="0"
            value="0"
          >
            page 1
          </Option>
          <Option
            key="1"
            value="1"
          >
            page 2
          </Option>
        </withChangeHandler(ItemSelect)>
      </Body>
    </PlainModal__Fieldset>
    <PositionSelectModal__StyledButtonGroup>
      <PositionSelectModal__StyledButton
        onClick={[Function]}
        secondary={true}
        type="button"
      >
        Cancel
      </PositionSelectModal__StyledButton>
      <PositionSelectModal__StyledButton
        primary={true}
        type="submit"
      >
        Move
      </PositionSelectModal__StyledButton>
    </PositionSelectModal__StyledButtonGroup>
  </form>
</PlainModal>
`;
